llama_index/objects/base_node_mapping.py

```diff
@@ -0,0 +1,68 @@
+"""Base object types."""
+
+from typing import TypeVar, Generic, Sequence, Type, Any
+from abc import abstractmethod
+from llama_index.data_structs.node import Node
+from llama_index.indices.base import BaseIndex
+from llama_index.indices.vector_store.base import VectorStoreIndex
+from llama_index.tools.types import BaseTool
+
+OT = TypeVar("OT")
+
+
+class BaseObjectNodeMapping(Generic[OT]):
+    """Base object node mapping."""
+
+    @classmethod
+    @abstractmethod
+    def from_objects(
+        cls, objs: Sequence[OT], *args: Any, **kwargs: Any
+    ) -> "BaseObjectNodeMapping":
+        """Initialize node mapping from a list of objects.
+
+        Only needs to be specified if the node mapping
+        needs to be initialized with a list of objects.
+
+        """
+        pass
+
+    @abstractmethod
+    def add_object(self, obj: OT) -> None:
+        """Add object.
+
+        Only needs to be specified if the node mapping
+        needs to be initialized with a list of objects.
+
+        """
+        pass
+
+    @abstractmethod
+    def to_node(self, obj: OT) -> OT:
+        """To node."""
+        pass
+
+    @abstractmethod
+    def from_node(self, node: Node) -> OT:
+        """From node."""
+
+
+class SimpleObjectNodeMapping(BaseObjectNodeMapping):
+    """General node mapping that works for any obj."""
```

Comment:
"... any obj with a meaningful string representation" ? Just trying to imagine the use-case for this one (SQL tables seem like a good option!)